wasm-runtime/polkadot/src/runtime/session.rs

```diff
@@ -17,111 +17,120 @@
-	ValidatorStorageVec::set_items(new);
-	consensus::set_authorities(new);
-}
-
 /// The number of blocks in each session.
 pub fn length() -> BlockNumber {
-	Storable::lookup_default(b"ses:len")
+	storage::get_or(b"ses:len", 0)
 }
 
-/// The current era index.
-pub fn current_index() -> BlockNumber {
-	Storable::lookup_default(b"ses:ind")
+/// The number of validators currently.
+pub fn validator_count() -> usize {
+	ValidatorStorageVec::count() as usize
 }
 
-/// Set the current era index.
-pub fn set_current_index(new: BlockNumber) {
-	new.store(b"ses:ind");
+/// The current era index.
+pub fn current_index() -> BlockNumber {
+	storage::get_or(b"ses:ind", 0)
 }
 
 /// The block number at which the era length last changed.
 pub fn last_length_change() -> BlockNumber {
-	Storable::lookup_default(b"ses:llc")
+	storage::get_or(b"ses:llc", 0)
 }
 
-/// Set a new era length. Won't kick in until the next era change (at current length).
-pub fn set_length(new: BlockNumber) {
-	new.store(b"ses:nln");
+pub mod public {
+	use super::*;
+
+	/// Sets the session key of `_validator` to `_key`. This doesn't take effect until the next
+	/// session.
+	pub fn set_key(validator: &AccountID, key: &SessionKey) {
+		// set new value for next session
+		storage::put(&validator.to_keyed_vec(b"ses:nxt:"), key);
+	}
+}
+
+pub mod privileged {
+	use super::*;
+
+	/// Set a new era length. Won't kick in until the next era change (at current length).
+	pub fn set_length(new: BlockNumber) {
+		storage::put(b"ses:nln", &new);
+	}
 }
 
-/// Hook to be called after transaction processing.
-pub fn check_rotate_session() {
-	// do this last, after the staking system has had chance to switch out the authorities for the
-	// new set.
-	// check block number and call next_session if necessary.
-	if (system::block_number() - last_length_change()) % length() == 0 {
-		rotate_session();
+// INTERNAL API (available to other runtime modules)
+
+pub mod internal {
+	use super::*;
+
+	/// Set the current set of validators.
+	///
+	/// Called by staking::next_era() only. `next_session` should be called after this in order to
+	/// update the session keys to the next validator set.
+	pub fn set_validators(new: &[AccountID]) {
+		ValidatorStorageVec::set_items(new);
+		consensus::internal::set_authorities(new);
+	}
+
+	/// Hook to be called after transaction processing.
+	pub fn check_rotate_session() {
+		// do this last, after the staking system has had chance to switch out the authorities for the
+		// new set.
+		// check block number and call next_session if necessary.
+		if (system::block_number() - last_length_change()) % length() == 0 {
+			rotate_session();
+		}
 	}
 }
 
-// PRIVATE (not available for use externally)
+struct ValidatorStorageVec {}
+impl StorageVec for ValidatorStorageVec {
+	type Item = AccountID;
+	const PREFIX: &'static[u8] = b"ses:val:";
+}
 
 /// Move onto next session: register the new authority set.
 fn rotate_session() {
```

Comment:
I think this should be a private API also, only called by `check_rotate_session`

Reply:
it is private - hence the lack of `pub`

Reply:
I mean that it is the only function which is not in a submodule. it's only referenced by `check_rotate_session` which is in the internal* module.